### ISSUE TITLE
Add Word2Vec Turkish vectorization

### ISSUE THREAD

**Issue by askarbozcan** [GlobalMaksimum/sadedegel#119]
Implement Word2Vec vectorizer as BERT might be too heavy for most usage cases and TF-IDF vectors might not be sufficient.

Might be a good idea to split this issue into multiple issues such as:
1) Add pre-trained Word2Vec model on Turkish
2) Implement necessary interface for Sadedegel

**Comment by askarbozcan** [comment]
Due to various ways to implement vectorization, with their own pros and cons below I am going to list some of the paths we can take for us to discuss. 

**1) Using pretrained fasttext model**
* **Pros**
  * fasttext as a word embedding model is more suitable for Turkish (unlike word2vec) due to the way it represents words.
For an example, the word `where` is represented as a bag of n-grams like so: `'<wh', 'whe', 'her', 'ere', 're>', '<where>'`
This means that for example `orada` and `oradaki` which under word2vec would be totally separate words, are better represented with a fasttext model. 
  * Another advantage of using fasttext is its ability to generate decent vectors for OOV tokens. Due to the way it represents words as it will decompose OOV into tokens in its vocabulary and take the average of those vectors to represent OOV token.
  * Integrating fasttext model into sadedegel would be a quite simple and fast ordeal
* **Cons**
  * Pretrained model size is **7.2GB** and it takes up **~4GB** of RAM which defeats the purpose of using word vectorization as a performant alternative to BERT.
  * Will need to add fasttext library as a dependency to Sadedegel.

**2) Using vector mapping supplied by fasttext developers**
Vector mapping is a file that maps a token to its vector embedding.
ex: `sade: [0.1,0.2, ..., 0.8]`

Vector mapping provided by fasttext devs has been produced by using tokens tokenized with [EuroParl tokenizer](https://www.statmt.org/europarl/) and works very similarly to our SimpleTokenizer with Turkish. (See fasttext vocab coverage at the end)

* **Pros**
  * Is ordered by frequency, so can just take most frequent 300k or so word vectors for reduced size on disk and in memory 
  * Won't take long to integrate into Sadedegel as a feature
  * No new dependencies

* **Cons**
  * Cannot utilize OOV nor bag of n-gram mechanisms of fasttext model
  * Is still pretty big in its fullest, at ~4GB uncompressed (2M tokens)
  * Takes around ~1GB in memory in its fullest

**3) Training fasttext model on our own corpus**
* **Pros**
  * Gives us the flexibility of using our own tokenization & vocabulary
  * Will work decently well even when trained on relatively small corpus (~2GB) due to OOV mechanisms listed above

* **Cons**
  * Need to build a new corpus, as even extended corpus is still too small to train a generic word embedding model
  * Need time to create a corpus for training.
  * Needs varied corpus (not just news) to cover wide variety of words and their relations
  * Dependence on fasttext library

**4) Use pretrained word2vec model**
* **Pros**
  * Pros of 1) with exception of dependence on gensim, which is a much larger library and undesired as a dependency.
* **Cons**
  * Again, very similar to 1)

**5) Train word2vec  on our own corpus**
* **Pros**
  * Can use Pytorch to implement the model, allowing us to add vectorization with no new dependencies
  * Flexibility of tokenization/corpus choice
* **Cons**
  * Need a new corpus, but unlike 3) we will need a larger corpus since in word2vec `orada` and `oradaki` for example are two completely different words (even if they would be extremely close in vector space).
  * No OOV mechanism, will be forced to use random/zero vector for OOV tokens.

**6) Use word2vec vector mapping**
* Very similar to 2).

**7) Custom grapheme-level based model**
To accomodate agglutinative language such as Turkish, a grapheme-level word embedding model can be implemented.
A possible reference: https://www.aclweb.org/anthology/L18-1471.pdf
* **Pros**
  * Specialized towards Turkish, so the model will possibly understand suffixes much better than f.e fasttext

* **Cons**
  * Needs corpus
  * By far the longest and most laborious option, might take atleast a month to R&D.

 
---
**How much of extended corpus' tokens is covered by fasttext vector mapping**
To check if fasttext vector mapping has a good coverage, the whole extended corpus has been tokenized with SimpleTokenizer.
Below are the percentages of tokens in extended corpus which are present in fasttext vector mapping:
* **full mapping** - %75
* **first 300k tokens** - %30

It is worth noting that in the case of tokens not covered by full mapping a lot of the tokens are words with a lot of suffixes, names of orgs/people or are non-Turkish words.

---

**Conclusion**
Given that the most feasible (in terms of quality/performance) option which is using first 300k tokens of fasttext vector mapping has a low coverage of tokens in extended corpus I feel that the next most feasible option would be to train a fasttext model on a new corpus with a smaller vocabulary and integrating fasttext's training for more advanced users of Sadedegel.
